skynet.go

```diff
@@ -57,33 +132,259 @@ type (
+		ConnectionOptions:      DefaultConnectionOptions,
+		PortalCreateSkykeyPath: "/skynet/createskykey",
+	}
+	// DefaultGetSkykeyIDOptions contains the default skykeyid GET options.
+	DefaultGetSkykeyIDOptions = GetSkykeyIDOptions{
+		ConnectionOptions:     DefaultConnectionOptions,
+		PortalGetSkykeyIDPath: "/skynet/skykeyid",
+	}
+	// DefaultGetSkykeyOptions contains the default skykey GET options.
+	DefaultGetSkykeyOptions = GetSkykeyOptions{
+		ConnectionOptions:   DefaultConnectionOptions,
+		PortalGetSkykeyPath: "/skynet/skykey",
+	}
+	// DefaultListSkykeysOptions contains the default skykeys options.
+	DefaultListSkykeysOptions = ListSkykeysOptions{
+		ConnectionOptions:     DefaultConnectionOptions,
+		PortalListSkykeysPath: "/skynet/skykeys",
+	}
+
 	// DefaultDownloadOptions contains the default download options.
 	DefaultDownloadOptions = DownloadOptions{
-		PortalURL:          "https://siasky.net",
+		ConnectionOptions:  DefaultConnectionOptions,
 		PortalDownloadPath: "/",
 	}
 
 	// DefaultUploadOptions contains the default upload options.
 	DefaultUploadOptions = UploadOptions{
-		PortalURL:                    "https://siasky.net",
+		ConnectionOptions:            DefaultConnectionOptions,
 		PortalUploadPath:             "/skynet/skyfile",
 		PortalFileFieldName:          "file",
 		PortalDirectoryFileFieldName: "files[]",
 	}
+
+	// ErrResponseError is the error for a response with a status code >= 400.
+	ErrResponseError = errors.New("error response")
 )
 
+// AddSkykey stores the given base-64 encoded skykey with the renter's skykey
+// manager.
+func AddSkykey(skykey string, opts AddSkykeyOptions) error {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalAddSkykeyPath)
+	url = fmt.Sprintf("%s?skykey=%s", url, skykey)
+
+	req, err := makeRequest(opts.ConnectionOptions, "POST", url, body)
+	if err != nil {
+		return errors.AddContext(err, "could not make request")
+	}
+
+	// Add the skykey.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.AddContext(err, "could not execute POST")
+	}
+	if resp.StatusCode >= 400 {
+		return errors.AddContext(makeResponseError(resp), "error code received")
+	}
+
+	return nil
+}
+
+// CreateSkykey returns a new skykey created and stored under the given name
+// with the given type. skykeyType can be either "public-id" or "private-id".
+func CreateSkykey(name, skykeyType string, opts CreateSkykeyOptions) (Skykey, error) {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalCreateSkykeyPath)
+	url = fmt.Sprintf("%s?name=%s&type=%s", url, name, skykeyType)
+
+	req, err := makeRequest(opts.ConnectionOptions, "POST", url, body)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not make request")
+	}
+
+	// Create the skykey.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not execute POST")
+	}
+	if resp.StatusCode >= 400 {
```

Comment:
might be worth in a follow up MR to create a helper methods that does the error handling and status code checking for all the methods instead of calling `client.Do`.